File: mini-cursor/minicursor/providers/openai_backend.py
"""Backend for any OpenAI-compatible chat-completions endpoint.

Covers OpenAI itself plus OpenRouter, Groq, DeepSeek, Mistral, xAI, and local
servers (Ollama, LM Studio, vLLM) — anything speaking /v1/chat/completions
with function tools and streaming.
"""

from __future__ import annotations

import json
import os

from ..config import LOCAL_HOSTS, ProviderConfig, default_context_window
from ..ui import UI
from .base import Backend, BackendConfigError, ToolCall, ToolResult, TurnResult


def to_openai_tools(tools: list[dict]) -> list[dict]:
    return [
        {
            "type": "function",
            "function": {
                "name": t["name"],
                "description": t["description"],
                "parameters": t["input_schema"],
            },
        }
        for t in tools
    ]


def _resolve_api_key(pcfg: ProviderConfig) -> str:
    if pcfg.api_key_env:
        key = os.environ.get(pcfg.api_key_env)
        if key:
            return key
    if pcfg.base_url and any(h in pcfg.base_url for h in LOCAL_HOSTS):
        return "local"  # local servers don't check the key
    hint = pcfg.api_key_env or "an api_key_env entry in ~/.minicursor/config.json"
    raise BackendConfigError(f"provider '{pcfg.name}' needs an API key — set {hint}")


class OpenAIBackend(Backend):
    def __init__(self, pcfg: ProviderConfig, client=None):
        super().__init__(pcfg.model, context_window=default_context_window(pcfg))
        self.name = pcfg.name
        if client is not None:
            self.client = client
        else:
            from openai import OpenAI

            self.client = OpenAI(base_url=pcfg.base_url, api_key=_resolve_api_key(pcfg))
        # None = untested, True/False = known from a prior call. Some
        # OpenAI-compatible endpoints (notably some local servers) reject the
        # stream_options parameter outright, so we probe once and remember.
        self._usage_supported: bool | None = None

    def add_user_message(self, messages: list, text: str) -> None:
        messages.append({"role": "user", "content": text})

    def add_tool_results(self, messages: list, results: list[ToolResult]) -> None:
        for r in results:
            messages.append(
                {"role": "tool", "tool_call_id": r.call_id, "content": r.output}
            )

    def _create_stream(self, full_messages: list, tools: list[dict]):
        kwargs = dict(model=self.model, messages=full_messages, tools=to_openai_tools(tools), stream=True)
        if self._usage_supported is not False:
            try:
                stream = self.client.chat.completions.create(
                    **kwargs, stream_options={"include_usage": True}
                )
                self._usage_supported = True
                return stream
            except Exception:
                # Doesn't support stream_options (or some other one-off failure at
                # request time, before any output was produced) — fall back and
                # remember, so we don't pay for a failed request every call.
                self._usage_supported = False
        return self.client.chat.completions.create(**kwargs)

    def stream_turn(self, messages: list, system: str, tools: list[dict], ui: UI) -> TurnResult:
        stream = self._create_stream([{"role": "system", "content": system}] + messages, tools)

        text_parts: list[str] = []
        pending: dict[int, dict] = {}  # tool-call accumulator keyed by index
        finish_reason = None

        for chunk in stream:
            usage = getattr(chunk, "usage", None)
            if usage is not None:
                prompt_tokens = getattr(usage, "prompt_tokens", None)
                if prompt_tokens is not None:
                    self.last_context_tokens = prompt_tokens
            if not getattr(chunk, "choices", None):
                continue
            choice = chunk.choices[0]
            delta = choice.delta
            if delta and delta.content:
                text_parts.append(delta.content)
                ui.stream_text(delta.content)
            if delta and delta.tool_calls:
                for tc in delta.tool_calls:
                    slot = pending.setdefault(tc.index, {"id": "", "name": "", "args": ""})
                    if tc.id:
                        slot["id"] = tc.id
                    if tc.function:
                        if tc.function.name:
                            slot["name"] = tc.function.name
                        if tc.function.arguments:
                            slot["args"] += tc.function.arguments
            if choice.finish_reason:
                finish_reason = choice.finish_reason
        ui.flush_stream()

        assistant: dict = {"role": "assistant", "content": "".join(text_parts) or None}
        calls: list[ToolCall] = []
        if pending:
            assistant["tool_calls"] = []
            for i, slot in sorted(pending.items()):
                call_id = slot["id"] or f"call_{i}"
                assistant["tool_calls"].append(
                    {
                        "id": call_id,
                        "type": "function",
                        "function": {"name": slot["name"], "arguments": slot["args"]},
                    }
                )
                try:
                    parsed = json.loads(slot["args"]) if slot["args"].strip() else {}
                except json.JSONDecodeError:
                    parsed = {}
                calls.append(ToolCall(id=call_id, name=slot["name"], input=parsed))
        messages.append(assistant)

        if calls:
            return TurnResult("tool_use", calls)
        if finish_reason == "length":
            return TurnResult("length")
        return TurnResult("end")

    # -- context-compaction support ---------------------------------------

    def render_round(self, round_messages: list) -> str:
        lines: list[str] = []
        for m in round_messages:
            role = m.get("role", "?")
            if role == "tool":
                lines.append(f"tool result: {str(m.get('content', ''))[:1000]}")
                continue
            content = m.get("content")
            if isinstance(content, str) and content:
                lines.append(f"{role}: {content}")
            for call in m.get("tool_calls") or []:
                fn = call.get("function", {}) if isinstance(call, dict) else {}
                name = fn.get("name")
                args = str(fn.get("arguments", ""))[:500]
                lines.append(f"{role} called tool {name}({args})")
        return "\n".join(lines)

    def complete_text(self, system: str, user_text: str) -> str:
        response = self.client.chat.completions.create(
            model=self.model,
            messages=[{"role": "system", "content": system}, {"role": "user", "content": user_text}],
        )
        return response.choices[0].message.content or ""

File: tythan-code/src/core/providers/openaiBackend.ts
/**
 * Backend for any OpenAI-compatible chat-completions endpoint. Covers OpenAI
 * itself plus OpenRouter, Groq, DeepSeek, Mistral, xAI, and local servers
 * (Ollama, LM Studio, vLLM). Ported from the Python mini-cursor CLI's
 * openai_backend.py, including the fix found reviewing it: the
 * stream_options-support probe only treats a request-shape rejection
 * (BadRequestError / UnprocessableEntityError / TypeError) as "unsupported",
 * so an unrelated auth/rate-limit/network failure isn't misclassified and
 * doesn't permanently disable usage tracking.
 */

import OpenAI, { BadRequestError, UnprocessableEntityError } from "openai";
import type { ChatCompletionChunk, ChatCompletionMessageParam, ChatCompletionTool } from "openai/resources/chat/completions";

import { LOCAL_HOSTS, defaultContextWindow } from "../config";
import type { ProviderConfig } from "../config";
import type { NativeMessage } from "../compaction";
import { truncate } from "../tools";
import type { ToolDefinition } from "../tools";
import { BackendConfigError, Backend, makeTurnResult } from "./types";
import type { StreamSink, ToolCallRequest, ToolResultInput, TurnResult } from "./types";

export function toOpenAiTools(tools: ToolDefinition[]): ChatCompletionTool[] {
  return tools.map((t) => ({
    type: "function" as const,
    function: {
      name: t.name,
      description: t.description,
      parameters: t.inputSchema,
    },
  }));
}

export function resolveApiKey(pcfg: ProviderConfig): string {
  if (pcfg.apiKey) {
    return pcfg.apiKey;
  }
  if (pcfg.baseUrl && LOCAL_HOSTS.some((h) => pcfg.baseUrl?.includes(h))) {
    return "local"; // local servers don't check the key
  }
  throw new BackendConfigError(`provider '${pcfg.name}' needs an API key`);
}

interface PendingToolCall {
  id: string;
  name: string;
  args: string;
}

export class OpenAIBackend extends Backend {
  readonly name: string;
  private readonly client: OpenAI;
  /** undefined = untested, true/false = known from a prior call. Some
   * OpenAI-compatible endpoints (notably some local servers) reject the
   * stream_options parameter outright, so we probe once and remember. */
  private usageSupported: boolean | undefined = undefined;

  constructor(pcfg: ProviderConfig, client?: OpenAI) {
    super(pcfg.model, defaultContextWindow(pcfg));
    this.name = pcfg.name;
    this.client = client ?? new OpenAI({ baseURL: pcfg.baseUrl, apiKey: resolveApiKey(pcfg) });
  }

  addUserMessage(messages: NativeMessage[], text: string): void {
    messages.push({ role: "user", content: text });
  }

  addToolResults(messages: NativeMessage[], results: ToolResultInput[]): void {
    for (const r of results) {
      messages.push({ role: "tool", tool_call_id: r.callId, content: r.output });
    }
  }

  private async createStream(
    fullMessages: ChatCompletionMessageParam[],
    tools: ToolDefinition[],
  ): Promise<AsyncIterable<ChatCompletionChunk>> {
    const base = { model: this.model, messages: fullMessages, tools: toOpenAiTools(tools), stream: true as const };
    if (this.usageSupported === false) {
      return this.client.chat.completions.create(base);
    }
    try {
      const stream = await this.client.chat.completions.create({
        ...base,
        stream_options: { include_usage: true },
      });
      this.usageSupported = true;
      return stream;
    } catch (err) {
      if (err instanceof BadRequestError || err instanceof UnprocessableEntityError || err instanceof TypeError) {
        // These specifically mean "the request body/params were rejected" —
        // the closest thing to a reliable signal that stream_options isn't
        // recognized. Remember it and fall back so we don't pay for a failed
        // request every call. Anything else (auth, rate limit, network, 5xx)
        // is unrelated to stream_options support and must propagate normally
        // instead of permanently disabling usage tracking for an unrelated,
        // possibly transient reason.
        this.usageSupported = false;
        return this.client.chat.completions.create(base);
      }
      throw err;
    }
  }

  async streamTurn(
    messages: NativeMessage[],
    system: string,
    tools: ToolDefinition[],
    sink: StreamSink,
  ): Promise<TurnResult> {
    const fullMessages = [{ role: "system" as const, content: system }, ...messages] as ChatCompletionMessageParam[];
    const stream = await this.createStream(fullMessages, tools);

    const textParts: string[] = [];
    const pending = new Map<number, PendingToolCall>();
    let finishReason: string | null = null;

    for await (const chunk of stream) {
      if (chunk.usage) {
        const promptTokens = chunk.usage.prompt_tokens;
        if (promptTokens !== undefined) {
          this.lastContextTokens = promptTokens;
        }
      }
      const choice = chunk.choices?.[0];
      if (!choice) {
        continue;
      }
      const delta = choice.delta;
      if (delta?.content) {
        textParts.push(delta.content);
        sink.streamText(delta.content);
      }
      if (delta?.tool_calls) {
        for (const tc of delta.tool_calls) {
          const slot = pending.get(tc.index) ?? { id: "", name: "", args: "" };
          if (tc.id) {
            slot.id = tc.id;
          }
          if (tc.function?.name) {
            slot.name += tc.function.name;
          }
          if (tc.function?.arguments) {
            slot.args += tc.function.arguments;
          }
          pending.set(tc.index, slot);
        }
      }
      if (choice.finish_reason) {
        finishReason = choice.finish_reason;
      }
    }
    sink.flushStream();

    const assistant: NativeMessage = { role: "assistant", content: textParts.join("") || null };
    const calls: ToolCallRequest[] = [];
    if (pending.size > 0) {
      const toolCallsForHistory: unknown[] = [];
      for (const [index, slot] of [...pending.entries()].sort((a, b) => a[0] - b[0])) {
        const callId = slot.id || `call_${index}`;
        toolCallsForHistory.push({
          id: callId,
          type: "function",
          function: { name: slot.name, arguments: slot.args },
        });
        let parsed: Record<string, unknown> = {};
        try {
          parsed = slot.args.trim() ? (JSON.parse(slot.args) as Record<string, unknown>) : {};
        } catch {
          parsed = {};
        }
        calls.push({ id: callId, name: slot.name, input: parsed });
      }
      assistant.tool_calls = toolCallsForHistory;
    }
    messages.push(assistant);

    if (calls.length > 0) {
      return makeTurnResult("tool_use", calls);
    }
    if (finishReason === "length") {
      return makeTurnResult("length");
    }
    return makeTurnResult("end");
  }

  // -- context-compaction support ---------------------------------------

  override renderRound(roundMessages: NativeMessage[]): string {
    const lines: string[] = [];
    for (const m of roundMessages) {
      const role = m.role ?? "?";
      if (role === "tool") {
        lines.push(`tool result: ${truncate(String(m.content ?? ""), 1000)}`);
        continue;
      }
      const content = m.content;
      if (typeof content === "string" && content) {
        lines.push(`${role}: ${content}`);
      }
      const toolCalls = (m.tool_calls as Array<Record<string, unknown>> | undefined) ?? [];
      for (const call of toolCalls) {
        const fn = (call.function as Record<string, unknown> | undefined) ?? {};
        const args = truncate(String(fn.arguments ?? ""), 500);
        lines.push(`${role} called tool ${String(fn.name)}(${args})`);
      }
    }
    return lines.join("\n");
  }

  override async completeText(system: string, userText: string): Promise<string> {
    // Capped like AnthropicBackend.completeText — this is only ever used for
    // a compaction summary, so a verbose/non-compliant model can't return an
    // unbounded completion that defeats the point of compacting.
    const response = await this.client.chat.completions.create({
      model: this.model,
      max_tokens: 2000,
      messages: [
        { role: "system", content: system },
        { role: "user", content: userText },
      ],
    });
    return response.choices[0]?.message.content ?? "";
  }
}
